app/controllers/api/v1/bookmarks_controller.rb

```diff
@@ -0,0 +1,117 @@
+class Api::V1::BookmarksController < Api::V1::BaseController
+  respond_to :json
+
+  def create
+    archivist = User.find_by_login(params[:archivist])
+    external_bookmarks = params[:bookmarks]
+    bookmarks_responses = []
+    @bookmarks = []
+
+    # check for top-level errors (not an archivist, no bookmarks...)
+    status, messages = batch_errors(archivist, external_bookmarks)
+
+    if status == :ok
+      # Flag error and successes
+      @some_errors = @some_success = false
+
+      # Process the works, updating the flags
+      external_bookmarks.each do |external_bookmark|
+        bookmarks_responses << import_bookmark(archivist, external_bookmark)
+      end
+
+      # set final response code and message depending on the flags
+      messages = response_message(messages)
+    end
+
+    render status: status, json: { status: status, messages: messages, bookmarks: bookmarks_responses }
+  end
+
+  private
+
+  # Set messages based on success and error flags
+  def response_message(messages)
+    if @some_success && @some_errors
+      messages << "At least one bookmark was not created. Please check the bookmark array for further information."
```

Comment:
Good point - that's an implementation detail that doesn't need to be in the message. The developer who implements the API will presumably turn the JSON array into something more user friendly, so I'll rephrase that as "individual bookmark messages".